File: backend/src/settings/settings.controller.ts
import { Controller, Get, Put, Post, Body, UseGuards, Inject, forwardRef, BadRequestException } from '@nestjs/common';
import { SettingsService } from './settings.service';
import { MailService } from '../mail/mail.service';
import { TeamsService } from '../notifications/teams.service';
import { UpdateAuthSettingsDto } from './dto/update-auth-settings.dto';
import { UpdateEmailSettingsDto } from './dto/update-email-settings.dto';
import { UpdateAlertSettingsDto } from './dto/update-alert-settings.dto';
import { Roles } from '../auth/decorators/roles.decorator';
import { UserRole } from '../auth/enums/user-role.enum';
import { RolesGuard } from '../auth/guards/roles.guard';
import { Public } from '../auth/decorators/public.decorator';

@Controller('settings')
export class SettingsController {
  constructor(
    private readonly settingsService: SettingsService,
    @Inject(forwardRef(() => MailService))
    private readonly mailService: MailService,
    @Inject(forwardRef(() => TeamsService))
    private readonly teamsService: TeamsService,
  ) {}

  @Public()
  @Get('auth')
  getAuthSettings() {
    return this.settingsService.getAuthSettings();
  }

  @Put('auth')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  updateAuthSettings(@Body() updateDto: UpdateAuthSettingsDto) {
    return this.settingsService.updateAuthSettings(updateDto);
  }

  @Get('email')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  getEmailSettings() {
    return this.settingsService.getEmailSettings();
  }

  @Put('email')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  updateEmailSettings(@Body() updateDto: UpdateEmailSettingsDto) {
    return this.settingsService.updateEmailSettings(updateDto);
  }

  @Get('alerting')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  getAlertSettings() {
    return this.settingsService.getAlertSettings();
  }

  @Put('alerting')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  updateAlertSettings(@Body() updateDto: UpdateAlertSettingsDto) {
    return this.settingsService.updateAlertSettings(updateDto);
  }

  @Post('email/test')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  async sendTestEmail(@Body('email') email: string) {
    try {
      console.log(`[DEBUG] Attempting to send test email to: ${email}`);
      if (!this.mailService) {
        throw new Error('MailService is undefined in SettingsController');
      }
      await this.mailService.sendTestEmail(email);
      return { message: 'Email de test envoyé avec succès' };
    } catch (error) {
      console.error(`[ERROR] sendTestEmail failed: ${error.message}`);
      throw new BadRequestException(error.message);
    }
  }

  @Get('teams')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  getTeamsSettings() {
    return this.settingsService.getTeamsSettings();
  }

  @Put('teams')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  updateTeamsSettings(@Body() updateDto: { webhookUrl?: string, enabled?: boolean }) {
    return this.settingsService.updateTeamsSettings(updateDto);
  }

  @Post('teams/test')
  @Roles(UserRole.ADMIN, UserRole.SUPERUSER)
  @UseGuards(RolesGuard)
  async sendTeamsTest(@Body('webhookUrl') webhookUrl: string) {
    return this.teamsService.sendTestMessage(webhookUrl);
  }
}

File: backend-java/src/main/java/com/asmolabs/zanshin/settings/controllers/SettingsController.java
package com.asmolabs.zanshin.settings.controllers;

import com.asmolabs.zanshin.mail.services.MailService;
import com.asmolabs.zanshin.notifications.services.TeamsService;
import com.asmolabs.zanshin.settings.dto.UpdateAlertSettingsDto;
import com.asmolabs.zanshin.settings.dto.UpdateAuthSettingsDto;
import com.asmolabs.zanshin.settings.dto.UpdateEmailSettingsDto;
import com.asmolabs.zanshin.settings.services.SettingsService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/settings")
@RequiredArgsConstructor
public class SettingsController {

    private final SettingsService settingsService;
    private final MailService mailService;
    private final TeamsService teamsService;

    @GetMapping("/auth")
    public ResponseEntity<Map<String, Boolean>> getAuthSettings() {
        return ResponseEntity.ok(settingsService.getAuthSettings());
    }

    @PutMapping("/auth")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, Boolean>> updateAuthSettings(@RequestBody UpdateAuthSettingsDto updateDto) {
        // In a real app, we would get the user ID from the SecurityContext
        return ResponseEntity.ok(settingsService.updateAuthSettings(updateDto, "system"));
    }

    @GetMapping("/email")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, String>> getEmailSettings() {
        return ResponseEntity.ok(settingsService.getEmailSettings());
    }

    @PutMapping("/email")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, String>> updateEmailSettings(@RequestBody UpdateEmailSettingsDto updateDto) {
        return ResponseEntity.ok(settingsService.updateEmailSettings(updateDto, "system"));
    }

    @GetMapping("/alerting")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, String>> getAlertSettings() {
        return ResponseEntity.ok(settingsService.getAlertSettings());
    }

    @PutMapping("/alerting")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, String>> updateAlertSettings(@RequestBody UpdateAlertSettingsDto updateDto) {
        return ResponseEntity.ok(settingsService.updateAlertSettings(updateDto, "system"));
    }

    @PostMapping("/email/test")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, String>> sendTestEmail(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        mailService.sendTestEmail(email);
        return ResponseEntity.ok(Map.of("message", "Email de test envoyé avec succès"));
    }

    @GetMapping("/teams")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, Object>> getTeamsSettings() {
        return ResponseEntity.ok(settingsService.getTeamsSettings());
    }

    @PutMapping("/teams")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, Object>> updateTeamsSettings(@RequestBody Map<String, Object> updateDto) {
        return ResponseEntity.ok(settingsService.updateTeamsSettings(updateDto, "system"));
    }

    @PostMapping("/teams/test")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERUSER')")
    public ResponseEntity<Map<String, Object>> sendTeamsTest(@RequestBody Map<String, String> body) {
        String webhookUrl = body.get("webhookUrl");
        return ResponseEntity.ok(teamsService.sendTestMessage(webhookUrl));
    }
}
